### PR TITLE
fix(/lib/components/table): prevent scrollbars around `<Table>`s

### LINKED ISSUE #605: Scrollbars appear around tables when resizing the window or when table content is modified
**Describe the bug**
Quite often, tables are dynamic and rows are added, removed, hidden, shown, etc. Since the `<div>` that wraps the `<table>` has the class `overflow-x-auto`, this creates scrollbars any time the content of the table is modified. This also happens when resizing the window.

I suggest either removing `overflow-x-auto sm:rounded-lg` from the wrapper element, or making the rounded corners optional with a note in the docs that it is not recommended for tables that will be modified dynamically.

**To Reproduce**
Add any `<Table />` component then add or remove a row dynamically.

**Expected behavior**
No scrollbars should appear

**Project information:**

- Tailwind: 3.2.6
- Flowbite: 1.6.3
- Flowbite React: 0.3.8
- Type: Nextjs

### PR DESCRIPTION
**Pull request by multiwebinc**
## Description

This moves the background color and border radius to the cells and places a div under the table for the box shadow effect. That way if a table is modified (rows or columns added or removed) or if the screen resolution changes, it won't create scrollbars around the table.

Fixes #605

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Just tested by trying it out

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by tulup-conner** [comment]
I love this idea, but we don't want to ship custom CSS. Can you find a way to accomplish this in Tailwind CSS via the theme instead of creating a stylesheet file?

**Comment by multiwebinc** [comment]
@tulup-conner What about like this?

**Comment by tulup-conner** [comment]
We will want a solution using Tailwind CSS purely, which has support for custom selectors like these. Inline CSS has other negative implications for users, e.g., it will break Content Security Policies.

File: src/lib/components/Table/Table.tsx [hunk 1]
```diff
@@ -34,10 +35,30 @@ const TableComponent: FC<TableProps> = ({ children, striped, hoverable, classNam
   return (
     <div data-testid="table-element" className={classNames(theme.wrapper)}>
       <TableContext.Provider value={{ striped, hoverable }}>
+        <div className={classNames(theme.shadow, className)}></div>
         <table className={classNames(theme.base, className)} {...props}>
           {children}
         </table>
       </TableContext.Provider>
+      <style jsx>{`
```
Comment (rluders) [inline src/lib/components/Table/Table.tsx:43]: I don't think that we want to do this. You must use TailwindCSS classes. Flowbite is a TailwindCSS-based component library, there is no point to use CSS or SASS or whatever.

**Review by rluders — CHANGES_REQUESTED** [review]

File: src/lib/components/Table/Table.tsx (hunk 1 above)
Comment (rluders) [reply src/lib/components/Table/Table.tsx:43]: https://tailwindcss.com/docs/utility-first

**Comment by multiwebinc** [comment]
@rluders @tulup-conner I was looking into the [`group-{modifier}` class](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state), however I need 2 levels of groups, because it needs to select the first/last child of `<table>` combined with the first/last `<td>` or `<th>`. Would you be opposed to using https://www.npmjs.com/package/tailwindcss-scoped-groups? 

If not, the only thing I can do with tailwind is style the top borders in `<thead>` (if there is one) and bottom borders of `<tfoot>` (if there is one). Without knowing whether `<tbody>` is the first or last child, you can't add a border radius to any of the cells in it. Unless maybe I add two more props to the `<Table>` component - `headless` and `footless` and if either of these are `true`, it will add the border radius to the corresponding tbody cells.

**Comment by multiwebinc** [comment]
@rluders @tulup-conner Ok, I think that's it. No stylesheets. I didn't know about nested groups, so I didn't think it was possible without stylesheets.

**Comment by rluders** [comment]
@multiwebinc could you sync your branch with the `main` branch? It has some conflicts because the theme structure was changed. So, you may have to adjust a few things on your PR.

**Comment by multiwebinc** [comment]
@rluders What do you think, would it make sense to move the `table.cell` interface to `table.body.cell` to avoid confusion with `table.head.cell`?

**Comment by rluders** [comment]
> @rluders What do you think, would it make sense to move the `table.cell` interface to `table.body.cell` to avoid confusion with `table.head.cell`?

I think that it makes sense. WDYT, @tulup-conner ?

**Comment by rluders** [comment]
Well, I can't see any reason to keep this one pending. I'm accepting it, @multiwebinc. Thank you!

**Pull request merged**